File: leet_BST_108_sorted_array_to_btree___2pointer.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */

object Solution {
    def sortedArrayToBST(nums: Array[Int]): TreeNode = {
        if (nums == null)       return null
        if (nums.length == 0)   return null
        make_tree(nums, 0, nums.length - 1)
    }

    def make_tree(n: Array[Int], L: Int, R: Int) : TreeNode = {
        if (L > R)  null
        else {
            val mid = ((R - L) >> 1) + L
            // val mid = (R - L) / 2 + L
            val node = new TreeNode(n(mid))
            node.left = make_tree(n, L, mid - 1)
            node.right = make_tree(n, mid + 1, R)
            node
        }
    }
}

File: leet_BST_108_sorted_array_to_btree___2pointer.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums)
    {
        int n = (int) nums.size();
        if (!n)
            return NULL;
        if (n == 1)
            return new TreeNode(nums[0]);
        return make_tree(nums, 0, n - 1);
    }

    TreeNode*   make_tree(vector<int> nums, int L, int R)
    {
        TreeNode    *node;
        int         mid;

        if (L > R)
            return NULL;

        mid = (R - L) / 2 + L;
        node = new TreeNode(nums[mid]);
        node->left = make_tree(nums, L, mid - 1);
        node->right = make_tree(nums, mid + 1, R);
        return node ;
    }
};
